### PR TITLE
(VANAGON-40) Environmental Activism

### PR DESCRIPTION
**Pull request by mckern**
I'm trying something a little different for this feature addition... a "@rick used to work at github" style checklist and a longish-lived PR allowing others to track the status of a feature as it's developed.

- [X] Support for environment variables as sane objects
- [x] Doc & Tests for env. var class
- [x] DSL methods to define env. vars per `Project`
- [x] DSL methods to define env. vars per `Platform`
- [x] DSL methods to define env. vars per `Component`
- [x] Doc & test for shiny new DSL methods

### CODE REVIEW

File: lib/vanagon/project.rb [hunk 1]
```diff
@@ -46,6 +106,9 @@ def initialize(name, platform)
       @requires = []
       @directories = []
       @settings = {}
+      # Environments are like Hashes but with specific constraints
+      # around their keys and values.
+      @environment = Vanagon::Environment.new
```
Comment (melissa) [inline lib/vanagon/project.rb]: I assume there are going to be two seperate environment objects during a build. One defined here in the project, and one defined in the platform. How are those two going to interact?

Comment (mckern) [inline lib/vanagon/project.rb]:
```make
SHELL = bash

# The way I'm thinking about the break down, a Project will have 
# support for an Environment object which will contain a bunch 
# of top-level env. vars. which would be set at this level.
export VANAGON_STATSD_HOST := 127.0.0.1
export VANAGON_STATSD_PORT := 8125
export VANAGON_TEST_STRING := a test string
export VANAGON_PROJECT_NAME := vanagon-example
# And then that will end up merged with a Platform's Environment object,
# and the resulting complete Environment will be merged in and printed
# at the top scope of the Makefile. These would potentially be defined
# in a macOS platform definition, and they'd override any existing
# values of the same name in the Project's Environment or be appended
# if they don't already exist.
export CPPFLAGS = -O3 -I/opt/puppetlabs/puppet/include -march=core2 -msse4
export CC = pl-gcc-6

.PHONY: inherited
inherited: export VANAGON_TARGET = inherited
inherited: example
	echo "the target is $(@)"
	echo "the compiler is $(CC)"
	echo "and the CFLAGS are $(CPPFLAGS)"
	echo -e "----\n"

# And any component specific exports would end up defined in the Make
# target, and would override what's passed in from the top-level scope.
.PHONY: example
example: export VANAGON_TARGET = example
example: export CC = clang
example: export CPPFLAGS = -O1 -wall
example:
	echo "the target is $(@)"
	echo 'The compiler is $(CC)'
	echo "and the CFLAGS are $(CPPFLAGS)"
	echo -e "----\n"
```

```
mckern@flexo example $ make inherited
echo "the target is example"
the target is example
echo 'The compiler is clang'
The compiler is clang
echo "and the CFLAGS are -O1 -wall"
and the CFLAGS are -O1 -wall
echo -e "----\n"
----

echo "the target is inherited"
the target is inherited
echo "the compiler is pl-gcc-6"
the compiler is pl-gcc-6
echo "and the CFLAGS are -O3 -I/opt/puppetlabs/puppet/include -march=core2 -msse4"
and the CFLAGS are -O3 -I/opt/puppetlabs/puppet/include -march=core2 -msse4
echo -e "----\n"
----

mckern@flexo example $
```

Comment (melissa) [inline lib/vanagon/project.rb]: 👍

**Comment by mckern** [comment]
This is looking really good so far:

```make
export SHELL := $(shell which bash)

tempdir := $(shell mktemp -d -p /var/tmp 2>/dev/null || mktemp -d -t 'tmp')
workdir := $(PWD)

export VANAGON_PROJECT_VERSION = 1.8.3.1.gc9145783
export VANAGON_PROJECT_VENDOR = Puppet Labs <info@puppetlabs.com>
export VANAGON_PROJECT_NAME = puppet-agent
export VANAGON_PLATFORM_NAME = el-7-x86_64
```

**Comment by mckern** [comment]
tumblr_n5b0rx6d2b1qfhhceo8_250

**Comment by mckern** [comment]
Great news. I've built Puppet-Agent for `el-7-x86_64` and `debian-8-armhf` and it was 🙌 ! This work should be tested against AIX and Solaris, and if someone (like @rick maybe?) wanted to take a stab at the ABS engine, that'd be 👌 .

**Comment by mckern** [comment]
Here's some [Makefiles from before the new Environment work](https://gist.github.com/mckern/e238d630ec86718d03bb1a3f9e66c57b) and some Makefiles rendered [after the new Environment work](https://gist.github.com/mckern/195b6fe2f1f82d5159a7da0407ed4ad8). Diffs pending.

**Comment by melissa** [comment]
@mckern did you want to add updates to the readme before this PR is merged?

File: resources/Makefile.erb [hunk 2]
```diff
@@ -1,15 +1,15 @@
-<% dirnames = get_directories.map {|d| d.path } %>
```
Comment (melissa) [inline resources/Makefile.erb]: Why did you remove this line? Don't we need to set `dirnames`?

Comment (melissa) [inline resources/Makefile.erb]: OH! I found it, nevermind

File: lib/vanagon/platform/dsl.rb [hunk 3]
```diff
@@ -51,6 +51,7 @@ def platform(name, &block)
                     end
 
         yield(self)
+        environment 'VANAGON_PLATFORM', platform_name.tr('-', '.')
```
Comment (melissa) [inline lib/vanagon/platform/dsl.rb]: should we remove the `-` at a later point in time, since I think that's only really needed when we send metrics. Would it be confusing to have this variable be in the format `el.7.x86_64` rather than `el-7-x86_64`?

**Comment by melissa** [comment]
@mckern do you have a makefile diff yet?

**Comment by mckern** [comment]
@melissa the diffs: https://gist.github.com/mckern/d54f16aae3879e2f7760b0decad21cfa

**Comment by mckern** [comment]
I think I am going to omit adding to the README in this PR. I'm working on a restructuring of the docs, and that might fit better there.

File: lib/vanagon/platform/dsl.rb (hunk 3 above)
Comment (mckern) [inline lib/vanagon/platform/dsl.rb]: Well, if we use this value in a metrics aggregation system, `el.7.x86_64` would translate out as:

```
mckern@flexo Downloads $ tree -F el
el/
└── 7/
    └── x86_64/
        ├── stat_name_1
        ├── stat_name_2
        ├── stat_name_3
        ├── stat_name_4
        ├── stat_name_5
        └── stat_name_6
```

**Comment by melissa** [comment]
can you test this with a windows makefile? I'm curious to see what the diff there would look like

**Comment by mckern** [comment]
@melissa all relevant gists have been updated with a Windows 12 Makefile & associated diff.

File: lib/vanagon/platform/dsl.rb (hunk 3 above)
Comment (melissa) [inline lib/vanagon/platform/dsl.rb]: Will this variable only be used for stats? Is there a possibility we'll use it for something else? I'm just concerned we've set the precedent for platform tags to have dashes, and changing that paradigm here is a little confusion. That being said, this is the introduction of a new variable, so it's not the worst. I'm not blocking on this, I just want to make sure it's a deliberate decision

Comment (mckern) [inline lib/vanagon/platform/dsl.rb]: This is intended to be used in metric rollup; nothing else has used this value inside the Makefile, so I'm not *super* worried about it. Open to leaving it alone, or even renaming this variable and using `VANAGON_PLATFORM` for the origina value. @underscorgan, @shrug, @branan feedback? Thoughts?

**Comment by mckern** [comment]
@branan @ScottGarman @McdonaldSeanp do you want to take a look at/chime in on this? Gonna fish for a merge otherwise.

**Comment by ScottGarman** [comment]
I've taken a cursory look at this and I'm generally in favor of it. Adding a section to the README file with some example use cases seems worthwhile. LGTM otherwise and I'd like to play with this when I have more time.

**Comment by shrug** [comment]
Add a changelog update?
